File: problems/23/solution2.ts
// Accepted: 82ms (87.38%), 56.67MB (11.46%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    function go(left: number = 0, right: number = lists.length): ListNode | null {
        const diff = right - left;
        if (diff === 0) {
            return null;
        }
        if (diff === 1) {
            return lists[left];
        }

        const mid = left + Math.floor(diff / 2);
        const mergedLeft = go(left, mid);
        const mergedRight = go(mid, right);

        return merge(mergedLeft, mergedRight);
    }

    function merge(left: ListNode | null, right: ListNode | null): ListNode | null {
        if (left === null) {
            return right;
        }
        if (right === null) {
            return left;
        }

        if (left.val < right.val) {
            left.next = merge(left.next, right);
            return left;
        } else {
            right.next = merge(left, right.next);
            return right;
        }
    }

    return go();
};

File: problems/23/solution2.py
# Accepted: 83ms (69.15%), 20.31MB (39.09%)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def go(lists: List[Optional[ListNode]], left: int = 0, right: int = len(lists)):
            diff = right - left
            if diff == 0:
                return None
            if diff == 1:
                return lists[left]

            mid = left + diff // 2
            mergedLeft = go(lists, left, mid)
            mergedRight = go(lists, mid, right)

            return merge(mergedLeft, mergedRight)

        def merge(left: Optional[ListNode], right: Optional[ListNode]):
            if left is None:
                return right
            if right is None:
                return left

            if left.val < right.val:
                left.next = merge(left.next, right)
                return left
            else:
                right.next = merge(left, right.next)
                return right

        return go(lists)
